common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,905 @@
+    String metaName = String.format("%s.meta", generateFileName(appShuffleId, reduceId));
+    return getFile(appShuffleId.appId, metaName);
+  }
+
+  @Override
+  public String[] getMergedBlockDirs(String appId) {
+    AppPathsInfo appPathsInfo = Preconditions.checkNotNull(appsPathInfo.get(appId),
+      "application " + appId + " is not registered or NM was restarted.");
+    String[] activeLocalDirs = Preconditions.checkNotNull(appsPathInfo.get(appId).activeLocalDirs,
+      "application " + appId
+      + " active local dirs list has not been updated by any executor registration");
+    return Arrays.stream(activeLocalDirs)
+      .map(dir -> dir + getRelativePath(appPathsInfo.user, appId))
+      .toArray(String[]::new);
+  }
+
+  @Override
+  public void applicationRemoved(String appId, boolean cleanupLocalDirs) {
+    logger.info("Application {} removed, cleanupLocalDirs = {}", appId, cleanupLocalDirs);
+    // TODO: Change the message when this service is able to handle NM restart
+    AppPathsInfo appPathsInfo = Preconditions.checkNotNull(appsPathInfo.remove(appId),
+      "application " + appId + " is not registered or NM was restarted.");
+    Iterator<Map.Entry<AppShuffleId, Map<Integer, AppShufflePartitionInfo>>> iterator =
+      partitions.entrySet().iterator();
+    while (iterator.hasNext()) {
+      Map.Entry<AppShuffleId, Map<Integer, AppShufflePartitionInfo>> entry = iterator.next();
+      AppShuffleId appShuffleId = entry.getKey();
+      if (appId.equals(appShuffleId.appId)) {
+        iterator.remove();
+        for (AppShufflePartitionInfo partitionInfo : entry.getValue().values()) {
+          try {
+            partitionInfo.closeAllFiles();
+          } catch (IOException e) {
+            logger.error("Error closing merged shuffle files for appId {} shuffleId {} reduceId {}",
+              appId, partitionInfo.appShuffleId.shuffleId, partitionInfo.reduceId);
+          }
+        }
+      }
+    }
+    if (cleanupLocalDirs) {
+      Path[] dirs = Arrays.stream(getActiveLocalDirs(appPathsInfo.activeLocalDirs))
+        .map(dir -> dir.resolve(getRelativePath(appPathsInfo.user, appId)))
+        .toArray(Path[]::new);
+      directoryCleaner.execute(() -> deleteExecutorDirs(dirs));
+    }
+  }
+
+  /**
+   * Synchronously delete local dirs, executed in a separate thread.
+   */
+  private void deleteExecutorDirs(Path[] dirs) {
+    for (Path localDir : dirs) {
+      try {
+        if (Files.exists(localDir)) {
+          JavaUtils.deleteRecursively(localDir.toFile());
+          logger.debug("Successfully cleaned up directory: {}", localDir);
+        }
+      } catch (Exception e) {
+        logger.error("Failed to delete directory: {}", localDir, e);
+      }
+    }
+  }
+
+  @Override
+  public StreamCallbackWithID receiveBlockDataAsStream(PushBlockStream msg) {
+    // Retrieve merged shuffle file metadata
+    String[] blockIdParts = msg.blockId.split("_");
```

Comment:
this seems out of place to me to do the parsing here.  Can we use the BlockId.apply function to create the type it is and then parse use some parsing functions from the blockid?

Reply:
`BlockId` is part of the core module and is not available in the network-shuffle module. The `ExternalBlockHandler` also parses the block ids like this.

Reply:
What about storing the shuffle ID, **map ID** (and not mapIndex as at line 302), reduce ID in separate member variables within the `PushBlockStream`? 
And of course with their real types: 
- shuffle ID: int
- map ID: long (and not int as in the line 302, you can check this at [ShuffleBlockId](https://github.com/apache/spark/blob/4b6847fdd5491f7858290468f718d033f54e4ff8/core/src/main/scala/org/apache/spark/storage/BlockId.scala#L58)])
- reduce ID: int

This way the parsing errors are detected at the `PushBlockStream` construction which is much closer to the root cause than this point.

Reply:
@attilapiros When the blocks are pushed, we still using the mapIndex in the message. This is the code on the push-side: [pushPreparation](https://github.com/linkedin/spark/blob/7478a3edff46b77d325d30bd952d6ba0a2f479ff/core/src/main/scala/org/apache/spark/shuffle/ShuffleWriter.scala#L411).
The merge service needs the mapIndex and not the mapIds.

Reply:
@attilapiros right now using map index is necessary to keep track of which shuffle block has been merged for a given shuffle partition tracked inside the bitmap.
If it's a deterministic stage, then we do not want to distinguish 2 attempts of the same map task as 2 separate blocks, which could lead to data duplication.
If it's a nondeterministic stage, then we will need to rerun the entire shuffle map stage and redo the merge.
This will be handled by SPARK-32923.

Reply:
Thanks, but in this case this is definitely not block ID which starts with "shuffle_", right?
Moreover passing this values as typed member variables( instead of a String where these values are encoded) is still a good idea. Do you agree?

Reply:
Yes, that's a valid point that we are calling it a blockId and it starts with `shuffle_` but in practice it is not that. Also It would be better if we don't encode these as a string.
However, I think I can take this up a follow-up which would immediately proceeds this change. The reason is then this will require changes  not only to `PushBlockStream` but also a lot of changes to `ExternalBlockPusherClient` and `OneForOneBlockPusher`.  I can create a jira if everyone agrees?
@attilapiros @Victsm

Reply:
As it has effect on the message itself I would prefer not to be postponed.

Reply:
@attilapiros I'm not sure I fully understand your question.
The `blockId` inside `PushBlockStream` is shuffle partition block Id.
This is the same as the the usage of `blockId` for fetching blocks:
https://github.com/apache/spark/blob/a744fea3be12f1a53ab553040b95da730210bc88/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java#L86-L92

As for using a typed message instead of a string to represent the block id, do you see the same issue in the existing block fetching API as well?

Reply:
> This is the same as the the usage of blockId for fetching blocks ...

If it is the same then it would be https://github.com/apache/spark/blob/4b6847fdd5491f7858290468f718d033f54e4ff8/core/src/main/scala/org/apache/spark/storage/BlockId.scala#L58

Right?

But there the second component is the map ID (with long type) and not the map index (int type) and those two has very different meanings.

Reply:
Just took another look on the latest code in master branch, and for block fetching API it is indeed doing the string parsing on the client side instead of on the server side.
Given that, I think this is a valid change to make so we are using separate shuffleId (int), mapIndex (int), and reduceId (int) in the `PushBlockStream` message.

Reply:
> But there the second component is the map ID (with long type) and not the map index (int type) and those two has very different meanings.

Yeah, that part I agree with. In the next patch for the shuffle write path change, where block push gets initiated, we are right now reusing the current `ShuffleBlockId` by passing an int mapIndex to the long mapId field, which is probably hacky.
I think in addition to making the `PushBlockStream` RPC changes so we are parsing the blockId String inside `OneForOneBlockPusher`, we should also create a new block ID type representing this type of blockId that corresponds to what ShuffleBlockId used to be.
Any suggestions for the name of this type of block?

Reply:
@tgravescs 

> It seems like it would have fairly small impact if you keep it to just the PUshblockStream message and the creation and consumption of it, but maybe I'm missing something?

The impact is small, limited to just where it is created (inside `ShuffleWriter`) and where it consumed.
I now agree that it's better to make this change, especially given that the block fetching API has made a similar change to parse the String into concrete types on the client side, so the server side does not need to parse.

> the other question is in the prepareBlockPushRequests should it really be using ShuffleBlockId, seems like it would be easy to just create a push based version of the block id but I would have to look more at the full version to see what all that impacts.

Indeed, reusing `ShuffleBlockId` inside `ShuffleWriter#prepareBlockPushRequests` is a bit hacky.
Any suggestions for the name of this push based version of the shuffle block id?

Reply:
I'd just say something like ShufflePushBlockId